### DISCUSSION TITLE
PID Controller Output and PWM

### DISCUSSION THREAD

**Discussion by WoodyloveBo** [bitcraze/discussions#1526] [Q&A]
Hello,

I recently learned that the control outputs (u1, u2, u3, u4) from the PID controller are closer to the actual PWM values sent to the motors rather than force or torque values. Because of this, I am interested in creating a controller where the outputs are in Newtons and Torques, which are then converted to PWM signals. But before proceeding, I would like to understand what the thrust, roll, pitch, and yaw values passed to the power distribution represent in the current PID controller. Additionally, I am curious about which code is responsible for converting these values into PWM signals.

Thank you.

**Comment by gemenerik** [comment]
I may not completely understand what you are asking for, please elaborate, but [this](https://www.bitcraze.io/documentation/system/platform/cf2-coordinate-system/) is the general coordinate system we use. It is the [controller](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/) that takes in setpoints and outputs PWM commands. We also have some [documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/pwm-to-thrust/) on the relationship between PWM and thrust. Also check out this relevant discussion: #1503

**Reply by WoodyloveBo** [reply]
Thank you for reply

Based on [this documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/), it seems that the control inputs obtained from both the PID position controller and the PID attitude controller are used to convert the thrust value and the angular velocities needed to maintain the desired attitude into PWM signals that are then sent to the motors. If that is the case, why are the roll and pitch values divided by 2 in [this code](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c#L92-L102)? Additionally, I do not fully understand why the values of m1, m2, m3, and m4 are divided in that way based on thrust, roll, pitch, and yaw.

**Reply by gemenerik** [reply]
> If that is the case, why are the roll and pitch values divided by 2 in [this code](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c#L92-L102)?

I’m not sure why the values are divided by two. However, it's important to realize that the values don't directly correspond to physical measurements. In PID control, you can adjust or scale the outputs in different ways to achieve the desired effect.

> Additionally, I do not fully understand why the values of m1, m2, m3, and m4 are divided in that way based on thrust, roll, pitch, and yaw.

This division is related to how roll, pitch, and yaw are controlled. To adjust pitch, for example, around the goal thrust, you decrease thrust on the front motors and increase it on the back motors to tilt the drone forward. The same concept applies to roll and yaw control.
